### PR TITLE
Wip fixed size buffer wrap rk

### PR DESCRIPTION
**Pull request by rkuhn**

### CODE REVIEW

**Comment by drewhk** [comment]
You missed other affected implementations. Do a search on "%" in Graph, there are 3 other affected stages.

File: akka-stream/src/main/scala/akka/stream/impl/Buffers.scala [hunk 1]
```diff
@@ -68,23 +68,23 @@ private[akka] object FixedSizeBuffer {
...
     def nonEmpty: Boolean = used != 0
 
     def enqueue(elem: T): Unit = {
-      put(writeIdx, elem)
+      put(writeIdx, elem, false)
       writeIdx += 1
     }
 
-    protected def toOffset(idx: Int): Int
+    protected def toOffset(idx: Long, maintenance: Boolean): Int
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/Buffers.scala]: WTF is `maintenance` and why didn't it deserve a comment? :)

File: akka-stream/src/main/scala/akka/stream/impl/Buffers.scala [hunk 2]
```diff
@@ -101,23 +101,30 @@ private[akka] object FixedSizeBuffer {
...
-      put(writeIdx, null.asInstanceOf[T])
+      put(writeIdx, null.asInstanceOf[T], false)
     }
   }
 
   private[akka] final class ModuloFixedSizeBuffer[T](_size: Int) extends FixedSizeBuffer[T](_size) {
-    override protected def toOffset(idx: Int): Int = idx % capacity
+    override protected def toOffset(idx: Long, maintenance: Boolean): Int = {
+      if (maintenance && readIdx > Int.MaxValue) {
+        val shift = Int.MaxValue - Int.MaxValue % capacity
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/Buffers.scala]: No comment about intent?

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/Buffers.scala]: Also, please use parenthesis for intentional priority.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 3]
```diff
@@ -611,21 +602,14 @@ final class Balance[T](val outputPorts: Int, waitForAllDownstreams: Boolean) ext
   override val shape: UniformFanOutShape[T, T] = UniformFanOutShape[T, T](in, out: _*)
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
-    private val pendingQueue = Array.ofDim[Outlet[T]](outputPorts)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Nice. Handrolled ad-hoc buffers using arrays is good to not have to proofread.

**Pull request merged**